lua/nvim_lsp/pyls.lua

```diff
@@ -0,0 +1,28 @@
+local skeleton = require 'nvim_lsp/skeleton'
+local util = require 'nvim_lsp/util'
+local lsp = vim.lsp
+
+local cwd = vim.loop.cwd()
+
+skeleton.pyls = {
+  default_config = {
+    cmd = {"pyls"};
+    filetypes = {"python"};
+    root_dir = function() return cwd end;
+    log_level = lsp.protocol.MessageType.Warning;
+  };
```

Comment:
We may want to include the default settings available so that users know what they can share. We could do this either via putting those default settings, or putting it in the docs below? https://github.com/neoclide/coc.nvim/wiki/Language-servers#python

Reply:
Oh, I haven't seen this. That's a lot...

Either way is fine with me; what would you prefer?

Reply:
I do want users to be aware of the options. I always find it annoying when I don't know what I can set, so I think we should provide whatever info we can. My only hesitation on including settings was if they end up breaking between versions, but I think that it's probably fine. Then again, settings specified by the language server act as a complete override, which may surprise users who have home dir configs.

Let's put it in the docs, and try to keep it to as few lines as possible. if you just collapse the larger arrays

          "jedi_references": {
            "enabled": true
          },

into

          "jedi_references": { "enabled": true },

It'll look fine. I'd also double check that all the options you can find are included on a best effort basis, because that coc.nvim doc could be outdated.

Reply:
Oh and try to convert the json syntax into lua syntax.

```
{commandPath = "",configurationSources = { "pycodestyle" },enable = true,plugins = {jedi_completion = {enabled = true},jedi_hover = {enabled = true},jedi_references = {enabled = true},jedi_signature_help = {enabled = true},jedi_symbols = {all_scopes = true,enabled = true},mccabe = {enabled = true,threshold = 15},preload = {enabled = true},pycodestyle = {enabled = true},pydocstyle = {enabled = false,match = "(?!test_).*\\.py",matchDir = "[^\\.].*"},pyflakes = {enabled = true},rope_completion = {enabled = true},yapf = {enabled = true}},trace = {server = "verbose"}}
```

here's a starter. Just put something like "available settings" and then finally add `settings = {};` as a place holder in the default_config so that users know that's where they can put settings.

Reply:
Wouldn't lua syntax be 
```lua
settings = {
       pyls = {
         enable = true;
         trace = { server = "verbose"; };
         commandPath = "";
         configurationSources = [ "pycodestyle" ];
         plugins = {
           jedi_completion = { "enabled" = true; };
           jedi_hover = { "enabled" = true; };
           jedi_references = { "enabled" = true; };    
           jedi_signature_help = { "enabled" = true; };
           jedi_symbols = {
             enabled = true;
             all_scopes = true;
           };
           mccabe = {
             enabled = true;
             threshold = 15;
           };
           preload = { enabled = true; };
           pycodestyle = { enabled = true; };
           pydocstyle = {
             enabled = false;
             match = "(?!test_).*\\.py";
             matchDir = "[^\\.].*";
           };
           pyflakes = { enabled = true; };
           rope_completion = { enabled = true; };
           yapf = { enabled = true; };
         };
       };
```

Reply:
I'd just dump that in the doc for easier cut&paste.

Reply:
Yeah, by syntax, I meant table syntax. I just meant to give you a quick starting point because I can quickly convert things, and I left indentation up to you :P 

Except you have some mistakes since that's not valid Lua for the `"enabled"` and the array style in `configurationSources`.

I literally evaluated `vim.fn.json_decode(...)` and pasted the result for you to hopefully save time.

Reply:
Nice ;) But doing it by hand, I learn something. (For example, not to forget half of them...)

I can't find a proper documentation for pyls :/

Reply:
@norcalli where exactly should I put that? in the docstring between ``[[ ]]``, or in the `default_config` below?